<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [1]:
import pandas as pd
import psycopg2
import requests
from io import StringIO

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [ ]:
# текст запроса
query_3_1 = f'''
    select count(*) from vacancies
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipython-input-1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [ ]:
# текст запроса
query_3_2 = f'''
    select count(*) from employers
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

/tmp/ipython-input-3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [ ]:
# текст запроса
query_3_3 = f'''
    select count(*) from areas
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

/tmp/ipython-input-928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [ ]:
# текст запроса
query_3_4 = f'''
    select count(*) from industries
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

/tmp/ipython-input-2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [ ]:
# выводы по предварительному анализу данных

#### Выводы по предварительному анализу данных

1. В базе данных содержится:
    - 49 197 вакансий,
    - 23 501 работодателя,
    - 1 362 региона,
    - 294 сферы деятельности.
Такой объем данных говорит о хлорошем покрытии рынка труда.

2. Объём данных достаточный для дальнейшего проведения детального и предметного анализа.

3. Соотношение ваканский и работадателей примерно 2 к 1. В таком случае распределение ваканский по работодателям неравномерно. А значит рынок формируетсяв основном за счет крупных работодатлей, которые публикуют большую часть вакансий.

4. Достаточно много регионов в базе данных, значит вероятно потребуется использовать агрегирующие функции в процессе анализа.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query_4_1 = f'''
    SELECT
        a.name AS area,
        COUNT(v.id) AS cnt
    FROM vacancies v
    JOIN areas a
        ON v.area_id = a.id
    GROUP BY a.name    -- группируем по имени региона
    ORDER BY cnt DESC  -- сортируем в порядке убывания
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df.head()

/tmp/ipython-input-1648501374.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''
    SELECT
        COUNT(id) AS cnt
    FROM vacancies
    WHERE
        salary_from IS NOT NULL  -- проверяем, что хотя бы из двух полей с зарплатой не пустое
    OR
        salary_to IS NOT NULL
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

/tmp/ipython-input-970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [ ]:
# текст запроса
query_4_3 = f'''
    SELECT
        ROUND(AVG(salary_from)) AS avg_salary_from,
        ROUND(AVG(salary_to)) AS avg_salary_to
    FROM vacancies
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

/tmp/ipython-input-2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''
    SELECT
        schedule,
        employment,
        COUNT(*) AS cnt
    FROM vacancies
    GROUP BY     -- группируем по типу рабочего графика и типу тудоустройства
        schedule,
        employment
    ORDER BY cnt DESC
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

/tmp/ipython-input-2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [ ]:
# текст запроса
query_4_5 = f'''
    SELECT
        experience,
        COUNT(*) AS cnt
    FROM vacancies
    GROUP BY experience -- группируем по опыту
    ORDER BY cnt ASC    -- сортируем в порядке возрастания
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

/tmp/ipython-input-2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [ ]:
# выводы по детальному анализу вакансий

#### Выводы по детальному анализу данных

1. Наибольшее количество ваканский представлено в крупных городах. Лидер - Москва, что соответствует реалиям: в крупном мегаполисе, экономическом центре и столице сконцентрирован центрация бизнеса, IT-инфраструктура, и как следствие, расположено бОльшее количество ваканский.

2. Только у приблизительно половины ваканский (24073 из 49 197) указана заработная плата (хотя бы одно поле вилки). Это показывает низкую прозрачность рынка, работодатели не предоставляют полную информацию о зарплате в вакансиях. Возможной причиной такой непрозрачности является то, что работодатели обсуждают зарплату только на переговорах, личном интервью кандидата.

3. Средние значения границ вилки: 71065 и 110537 рублей. Имеется существенная вариативность уровня зарплат.

4. Самый популярный вид сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях - это полный день	и полная занятость. На втором месте - удаленная работа и полная занятость. Что отражает тенденции современного рынка вакансий. Если будет дальнейший рост популярности удаленной работы, то прогнозируется увеличение вакансий в регионах при сохранении центров в столице.

5. Самый распространенный требуемый опыт работы, указанный в вакансиях - от 1 года до 3 лет. Приблизительно в 2 раза меньше - от 3 до 6 лет. В разы реже работодатели ищут сотрудников без опыта или с опытом более 6 лет. Таким образом, рынок ориентирован на уже подготовленных специалистов.

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [4]:
# текст запроса
query_5_1 = f'''
    WITH employer_vacancies AS (
        SELECT
            e.name AS employer,     -- работодатель
            COUNT(v.id) AS cnt      -- количество вакансий
        FROM employers e
        JOIN vacancies v
            ON e.id = v.employer_id
        GROUP BY e.name            -- группируем по названию работодателя
    ),
    ranked_employers AS (
        SELECT
            employer,
            cnt,
            DENSE_RANK() OVER (ORDER BY cnt DESC) AS rank  -- ранжируем по числу вакансий
        FROM employer_vacancies
    )
    SELECT
        employer,                  -- работодатель
        cnt,                       -- количество вакансий
        rank                       -- ранг (место) по количеству вакансий
    FROM ranked_employers
    WHERE rank IN (1, 5)            -- берем 1 и 5 место
    ORDER BY rank;                  -- сортируем по месту
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

/tmp/ipython-input-3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


employer   cnt  rank
0         Яндекс  1933     1
1  Газпром нефть   331     5

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [6]:
# текст запроса
query_5_2 = f'''
    WITH e AS (
        SELECT
            area AS area_id,          -- регион
            COUNT(*) AS employers_cnt -- количество работодателе
        FROM employers
        GROUP BY area
    ),
    v AS (
        SELECT
            area_id,                  -- регион
            COUNT(*) AS vacancies_cnt -- количество вакансий
        FROM vacancies
        GROUP BY area_id
    )
    SELECT
        a.name AS region, -- название региона
        COALESCE(e.employers_cnt, 0) AS employers_cnt, -- 0, если работодателей нет
        COALESCE(v.vacancies_cnt, 0) AS vacancies_cnt  -- 0, если вакансий нет
    FROM areas a
    LEFT JOIN e ON e.area_id = a.id -- объединяем с таблицами, описанными выше
    LEFT JOIN v ON v.area_id = a.id
    WHERE COALESCE(v.vacancies_cnt, 0) = 0 -- выбираем регионы без вакансий
    ORDER BY employers_cnt DESC -- сортируем по убыванию количества работодателей
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

/tmp/ipython-input-2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region  employers_cnt  \
0                                               Россия            410   
1                                            Казахстан            207   
2                                   Московская область             75   
3                                   Краснодарский край             19   
4                                   Ростовская область             18   
..                                                 ...            ...   
588                                             Гуково              0   
589                                                Нея              0   
590                                      Краснозерское              0   
591                                        Долгоруково              0   
592  Октябрьская (Крыловский район,Краснодарский край)              0   

     vacancies_cnt  
0                0  
1                0  
2                0  
3                0  
4                0  
..             ...  
588              0  
589              0  
590              0  
591              0  
592              0  

[593 rows x 3 columns]

 Среди регионов, в которых нет вакансий, наибольшее количество работодателей в регионе Россия

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [12]:
# текст запроса
query_5_3 = f'''
    SELECT
        e.name AS employer,              -- работодатель
        COUNT(DISTINCT v.area_id) AS area_cnt -- количество уникальных регионов, в которых работодатель публикует вакансии
    FROM employers e
    JOIN vacancies v
        ON e.id = v.employer_id
    GROUP BY
        e.id,
        e.name
    ORDER BY area_cnt DESC;  -- сортируем по убыванию количества регионов
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

/tmp/ipython-input-201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer  area_cnt
0                                      Яндекс       181
1                                  Ростелеком       152
2                                  Спецремонт       116
3                      Поляков Денис Иванович        88
4                                   ООО ЕФИН         71
...                                       ...       ...
14901                            НПП Авиатрон         1
14902              Центр дистанционных торгов         1
14903  Городские Телекоммуникационные Системы         1
14904                       Введенский, Отель         1
14905              Группа Компаний Автоимпорт         1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [14]:
# текст запроса
query_5_4 = f'''
    SELECT
        COUNT(*) AS employers_cnt   -- количество работодателей
    FROM employers e
    LEFT JOIN employers_industries i -- присоединяем таблицу сфер деятельности
        ON e.id = i.employer_id
    WHERE i.employer_id IS NULL   -- выбираем работодателей с не указанной сферой деятельности
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

/tmp/ipython-input-3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employers_cnt
0           8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [16]:
# текст запроса
query_5_5 = f'''
    WITH four_indstrs_comps AS (
        SELECT
            e.name AS name -- название компании
        FROM employers e
        JOIN employers_industries i  -- присоединяем таблицу сфер деятельности
            ON e.id = i.employer_id
        GROUP BY e.id, e.name
        HAVING COUNT(DISTINCT i.industry_id) = 4  -- берем компании с 4 уникальными сферами деятельности
    )
    SELECT
        name
    FROM four_indstrs_comps
    ORDER BY name  -- сортируем имена в алфавитном списке
    OFFSET 2 -- пропускаем 1 и 2, чтобы начать с 3
    LIMIT 1  -- оставляем компанию, которая на 3 месте
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

/tmp/ipython-input-388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [18]:
# текст запроса
query_5_6 = f'''
    SELECT
        COUNT(DISTINCT e.id) AS employers_cnt   -- количество уникальных работодателей
    FROM employers e
    JOIN employers_industries ei -- соединяем с  таблицей сфер деятельности
        ON e.id = ei.employer_id
    JOIN industries i
        ON ei.industry_id = i.id -- соединяем с таблицей справочника сфер деятельности
    WHERE i.name = 'Разработка программного обеспечения'   -- выбираем работодателей где указана данная сфера деятельности
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

/tmp/ipython-input-821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


employers_cnt
0           3553

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [30]:
# код для получения списка городов-милионников
url = "https://ru.wikipedia.org/wiki/Города-миллионеры_России"

headers = {
    "User-Agent": "Mozilla/5.0 (Colab; pandas read_html)"
}

resp = requests.get(url, headers=headers, timeout=30)
resp.raise_for_status()
resp.encoding = "utf-8"

html = resp.text
tables = pd.read_html(StringIO(html))

# Берём таблицу, где есть колонка "Город"
df = next(t for t in tables if "Город" in t.columns)

million_cities = (
    df["Город"]
    .astype(str)
    .str.replace(r"\[.*?\]", "", regex=True)
    .str.replace(r"\(.*?\)", "", regex=True)
    .str.strip()
    .tolist()
)

cities = ", ".join("'" + c.replace("'", "''") + "'" for c in million_cities)
print(cities)

'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'


In [31]:
# текст запроса
query_5_7 = f'''
    WITH yandex AS (
        SELECT
            a.name AS city,               -- название города
            COUNT(v.id) AS vacancies_cnt  -- количество вакансий Яндекса в городе
        FROM vacancies v
        JOIN employers e                  -- соединяем вакансии с таблицей работодателей
            ON v.employer_id = e.id
        JOIN areas a                      -- соединяем вакансии с таблицей регионов
            ON v.area_id = a.id
        WHERE e.name = 'Яндекс'           -- оставляем только для Яндекса
        AND a.name IN ({cities})      -- только города-миллионники
        GROUP BY a.name                   -- группируем по городам
    )

    SELECT
        city,  -- город
        vacancies_cnt    -- количество вакансий
    FROM yandex

    UNION ALL

    SELECT
        'Total' AS city, -- добавляем строку с общим количеством вакансий компании
        SUM(vacancies_cnt) AS vacancies_cnt
    FROM yandex

    ORDER BY vacancies_cnt ASC;   -- сортируем по возрастанию количества вакансий
'''

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

/tmp/ipython-input-1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


city  vacancies_cnt
0              Омск           21.0
1         Челябинск           22.0
2        Красноярск           23.0
3         Волгоград           24.0
4             Пермь           25.0
5            Казань           25.0
6    Ростов-на-Дону           25.0
7            Самара           26.0
8               Уфа           26.0
9         Краснодар           30.0
10          Воронеж           32.0
11      Новосибирск           35.0
12  Нижний Новгород           36.0
13     Екатеринбург           39.0
14  Санкт-Петербург           42.0
15           Москва           54.0
16            Total          485.0

***

In [ ]:
# выводы по анализу работодателей

#### Выводы по анализу работодателей

1. Топ-1 работодатель по количеству ваканский - Яндекс (1933 вакансий). На пятом месте находится Газпром нефть (331 вакансия). Видим, что у крупных работодателей соотвественно большой набор вакантных мест.

2. Среди регионов, в которых нет вакансий, наибольшее количество работодателей зафиксировано в регионе "Россия" (410 работодателей). Что говорит о том, что нет активного размещения вакансий, не все компании используют сервис hh.ru как основной канал для найма сотрудников. Либо это может говорить о низком спросе.

3. Наибольшее количество регионов, в которых работодатель публикует вакансии у компании Яндекс (181 регион). На втором месте по числу регионов Ростелеком (152), на третьем - Спецремонт (116). При этом указано много локальных работодателей, которые размещают вакансии только в одном регионе.
Можно сделать вывод о том, что крупные игроки рынка (крупные работодатели) лидируют и по географии вакансий, и по количеству вакансий, таким образом они занимают внушающую часть рынка.

4. У 8419 работодателей не указана сфера деятельности. Что говорит о предоставлении работодателем неполной информации в профиле компании. Отсутствие указанной сферы дейтельности может снижать качество аналитики.

5. Компания, находящаяся на третьем месте в алфавитном списке среди работодателей с четырьмя сферами деятельности — 2ГИС.

6. У 3553 работодателей указана сфера деятельности «Разработка программного обеспечения», это подтверждает высокую представленность IT-компаний на рынке вакансий.

7. У компании Яндекс всего 485 вакансий в городах-миллионниках. Вакансии распределены неравномерно, больше всего вакансий сосредоточено в столице (Москва) и северной столице (Санкт-Петербург).

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [33]:
# текст запроса
query_6_1 = f'''
    SELECT
        COUNT(*) AS cnt
    FROM vacancies
    WHERE LOWER(name) LIKE '%data%'
    OR LOWER(name) LIKE '%данн%';
'''

In [34]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

/tmp/ipython-input-366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [35]:
# текст запроса
query_6_2 = f'''
    SELECT
        COUNT(*) AS cnt
    FROM vacancies
    WHERE  -- условия отнесения вакансии к DS
    (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (
            LOWER(name) LIKE '%ml%'
            AND LOWER(name) NOT LIKE '%html%'  -- не нужно брать вакансии по HTML
        )
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
    AND
    (
        LOWER(name) LIKE '%junior%'
        OR experience = 'Нет опыта'
        OR employment = 'Стажировка'
    )
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

/tmp/ipython-input-2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [37]:
# текст запроса
query_6_3 = f'''
    SELECT
        COUNT(*) AS cnt
    FROM vacancies
    WHERE -- условия отнесения вакансии к DS
    (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (
            LOWER(name) LIKE '%ml%'
            AND LOWER(name) NOT LIKE '%html%'  -- не нужно брать вакансии по HTML
        )
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
    AND
    (
        LOWER(key_skills) LIKE '%sql%'       -- указан SQL
        OR LOWER(key_skills) LIKE '%postgres%' -- указан Postgres
    )
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

/tmp/ipython-input-3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


cnt
0  229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [39]:
# текст запроса
query_6_4 = f'''
    SELECT
        COUNT(*) AS cnt
    FROM vacancies
    WHERE -- условия отнесения вакансии к DS
    (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (
            LOWER(name) LIKE '%ml%'
            AND LOWER(name) NOT LIKE '%html%'  -- не нужно брать вакансии по HTML
        )
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
    AND
    (
        LOWER(key_skills) LIKE '%python%'  -- указан Python
    )
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

/tmp/ipython-input-4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


cnt
0  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [49]:
# текст запроса
query_6_5 = f'''
    SELECT
        AVG(ARRAY_LENGTH(STRING_TO_ARRAY(key_skills, E'\t'), 1)) AS avg_skills_count
    FROM vacancies
    WHERE
        key_skills IS NOT NULL
    AND  -- условия отнесения вакансии к DS
    (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (
            LOWER(name) LIKE '%ml%'
            AND LOWER(name) NOT LIKE '%html%'  -- не нужно брать вакансии по HTML
        )
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
'''

In [50]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

/tmp/ipython-input-3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg_skills_count
0          6.546778

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [57]:
# текст запроса
query_6_6 = f'''
    SELECT
        experience,   -- тип требуемого опыта работы
        ROUND(AVG(    -- средняя зарплата, округлённая до целого чисоа
            CASE
                -- если заполнены оба поля с зарплатой
                WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL
                THEN (salary_from + salary_to) / 2.0
                -- если заполнено только одно из полей
                ELSE COALESCE(salary_from, salary_to)
            END
        )) AS avg_salary
    FROM vacancies
    WHERE   -- Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой
    (
        salary_from IS NOT NULL
        OR salary_to IS NOT NULL
    )
    AND     -- условия отнесения вакансии к DS
    (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (
            LOWER(name) LIKE '%ml%'
            AND LOWER(name) NOT LIKE '%html%'  -- не нужно брать вакансии по HTML
        )
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
    GROUP BY experience  -- группируем по опыту работы
    ORDER BY
    CASE experience
        WHEN 'Нет опыта' THEN 1
        WHEN 'От 1 года до 3 лет' THEN 2
        WHEN 'От 3 до 6 лет' THEN 3
        WHEN 'Более 6 лет' THEN 4
        ELSE 5
    END;
'''

In [58]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

/tmp/ipython-input-519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    144233.0
2       От 3 до 6 лет    256454.0
3         Более 6 лет    157933.0

Дата-сайентист с опытом работы от 3 до 6 лет в среднем может рассчитывать на зарплату 256454.

In [59]:
connection.close()

***

In [ ]:
# выводы по предметному анализу

#### Выводы по предметному анализу

1. 1771 вакансий имеет отношение к данным. Это небольшая доля от общего числа вакансий в наборе данных.

2. Для начинающего дата-сайентиста (junior, нет опыта, ищет стажировку) есть 51 подходящая вакансия. Видим, что в DS малое количество предложений на начинающих специалистов. Войти в DS начинающим достаточно сложно, рынок ориентирован на опытных специалистов.

3. В 229 вакансиях для DS в качестве ключевого навыка указан SQL или Postgres. Навыки работы с базами данных встречаются в вакансиях и являются важными.

4. Python достаточно популярен в требованиях работодателей к DS. Количество вакансий, в которых в качестве ключевого навыка указан Python: 357. Можем сказать, что это ключевой и наиболее популярный навык для дата-саентистов.

5. В среднем в вакансиях для DS указывают ~6-7 ключевых навыков. работодатели указывают необходимый перечень требований.

6. Средняя зарплата для DS с опытом от 3 до 6 лет составляет примерно 256454 рублей. Заработная плата зависит от опыта кандидата.

7. Data Science - высокооплачеваемая профессия, однако порог входа достаточно высок.

## Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

1. Рынок труда не равномерный по регионам и по работодателям, вакансии сконцентрированы в больших городах и у крупных работодателей.

2. Рынок труда ориентирован на специалистов с опытом. Начинающим специалистам трудно войти в профессию, не имея опыта.

3. Присутствует непрозрачность вакансий на рынке - не все работодатели указывают вилку зарплаты и сферу деятельности.

4. Data Science имеет высокую ценность на рынке труда: мало вакансий, большой конкурс, много требований к навыкам, но зарплаты привлекательны и выше среднего. Порог вхождения высокий.

**Какие ещё интересные исследования можно провести?**

- Анализ вакансий с удаленной работой
- Прогнозирование зарплаты на основе навыков и опыта
- Анализ зависимости требований в зависимости от работодателя и региона